protocol/message.go

```diff
@@ -35,13 +35,18 @@ func NewErrorResponse(e ErrorCode) Response {
 }
 
 type DirectoryProof struct {
-	Type int
-	AP   *m.AuthenticationPath
-	STR  *m.SignedTreeRoot
-	TB   *m.TemporaryBinding `json:",omitempty"`
+	Type  int
+	AP    *m.AuthenticationPath
+	STR   *m.SignedTreeRoot
+	TB    *m.TemporaryBinding `json:",omitempty"`
+	Error string              `json:",omitempty"`
 }
 
 func NewRegistrationProof(ap *m.AuthenticationPath, str *m.SignedTreeRoot,
-	tb *m.TemporaryBinding) *DirectoryProof {
-	return &DirectoryProof{RegistrationType, ap, str, tb}
+	tb *m.TemporaryBinding, err ErrorCode) *DirectoryProof {
+	df := &DirectoryProof{RegistrationType, ap, str, tb, ""}
+	if err != Success {
+		df.Error = err.Error().Error()
+	}
+	return df
```

Comment:
I'm saying that above, you're doing,

``
return NewRegistrationProof(ap, d.LatestSTR(), nil, Success), Success
```

but if you `return df, err` here, you can do,

```
return NewRegistrationProof(ap, d.LatestSTR(), nil, Success)
```